File: src/ast.ts
export type TypeName = 'i32' | 'void';

export type Program = {
  consts: ConstDecl[];
  fns: FnDecl[];
};

export type ConstDecl = {
  name: string;
  type: TypeName;
  expr: Expr;
};

export type FnDecl = {
  exported: boolean;
  inlineHint: boolean;
  name: string;
  params: { name: string; type: TypeName }[];
  ret: TypeName;
  body: Stmt[];
};

export type Stmt =
  | { kind: 'let'; name: string; type: TypeName; expr: Expr }
  | { kind: 'assign'; name: string; expr: Expr }
  | { kind: 'if'; cond: Expr; then: Stmt[]; elseBranch: Stmt[] | null }
  | { kind: 'while'; cond: Expr; body: Stmt[] }
  | { kind: 'return'; expr: Expr | null }
  | { kind: 'expr'; expr: Expr };

export type BinOp =
  | 'add'
  | 'sub'
  | 'mul'
  | 'and'
  | 'or'
  | 'xor'
  | 'shl'
  | 'shr'
  | 'eq'
  | 'ne'
  | 'lt'
  | 'le'
  | 'gt'
  | 'ge';

export type Expr =
  | { kind: 'num'; value: number }
  | { kind: 'var'; name: string }
  | { kind: 'bin'; op: BinOp; left: Expr; right: Expr }
  | { kind: 'call'; name: string; args: Expr[] };

File: crates/wasmchi/src/ast.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub items: Vec<Item>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Item {
    ExportFn(Function),
    Fn(Function),
    ImportFn(ImportFn),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ImportFn {
    pub name: String,
    pub params: Vec<Param>,
    pub ret_ty: Type,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Function {
    pub name: String,
    pub params: Vec<Param>,
    pub ret_ty: Type,
    pub body: Vec<Stmt>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Param {
    pub name: String,
    pub ty: Type,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    I32,
    F64,
    Bool,
    String,
    Void,
    JsObj,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Stmt {
    Let { name: String, expr: Expr },
    Return(Expr),
    Print(Expr),
    Expr(Expr),
    If { cond: Expr, then_body: Vec<Stmt>, else_body: Option<Vec<Stmt>> },
    While { cond: Expr, body: Vec<Stmt> },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expr {
    Int(i32),
    Float(u64), // f64::to_bits()
    Bool(bool),
    Str(String),
    Var(String),

    // Calls
    Call { callee: String, args: Vec<Expr> },
    CallExpr { callee: Box<Expr>, args: Vec<Expr> },

    // Property access
    Dot { base: Box<Expr>, prop: String },

    Binary { op: BinOp, left: Box<Expr>, right: Box<Expr> },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinOp {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
    Ne,
    Lt,
    Le,
    Gt,
    Ge,
}
